src/logging/LazyLoggerFactory.ts

```diff
@@ -51,7 +52,7 @@ class TemporaryLoggerFactory implements LoggerFactory {
     }
     // Emit all buffered log messages
     for (const { logger, level, message } of this.buffer.splice(0, this.buffer.length)) {
-      logger.log(level, message);
+      logger.log(level, message, { isPrimary: cluster.isMaster, pid: process.pid });
```

Comment:
This logs the cluster information at the time of actually logging the message, not of the time it was created. So to be fully correct this information would have to be stored in the cached message.

Additionally, if these messages get created before clustering, but the switch happens after clustering these would be duplicated for every worker right?

We know this can't happen because the switch is configured to happen before workers are created though.

I'm thinking what makes most sense is perhaps having an initial check in the switch call to make sure we are (still) in the primary process, and throw an error if not.

Reply:
I do understand the first paragraph (the creation time of `message` is indeed earlier in time than the creation time of the metadata object).

I am not completely following with what you are trying to explain after that. What do you mean with _the switch_?

The way it works now (best use case is just starting with `-c config/file.json -w 2` in worker mode) we can look at the logs of the UnsecureWebSocket. They will get logged by the Primary process first, and then by each worker as they start.

Spawning of a worker does not happen from that moment on in the code where the forking happens. On `cluster.fork()` an entire new process spawns and it boots the whole server from the start.

Reply:
> What do you mean with _the switch_?

Calling this function (it's called `switch`). This class catches logs before the `LoggerInitializer` is called. Once that initializer triggers it calls the `switch` function here to send those messages to the initialized logger.

Some of my comments might stem from not fully understanding how the worker things work yes, so I probably should just start the server with `-w 2` and some extra logging to fully realize what happens where 😅

Reply:
I might be wrong, but I think it doesn't really matter. The purpose of the metadata addition in the logging statement, was to known from which process (pid and master/worker) the log call was coming. As far as I can see, this will always be the same process as where the `switch` method will be called from too. So it should be no problem to add the metadata at that point, I think.

![image](https://user-images.githubusercontent.com/1684858/169288136-3d3b77c4-41a1-4122-842f-964fa8af9d30.png)

The main process starts and reaches the `cluster.fork()` point. From there 2 new processes start and they follow the exact same boot procedure, thus fixing the metadata to their own process. (it doesn't even matter if the `switch` moment happens before or after the `process.fork()` call, I guess)

Reply:
Yes my understanding of how the clusters work was wrong. It indeed does not make a difference so this solution is fine.